Librerie/libpdel-0.6.1/pdel/http/servlet/http_servlet_tmpl.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: http_servlet_tmpl.3 901 2004-06-02 17:24:39Z archie $
.\"
.Dd April 22, 2002
.Dt HTTP_SERVLET_TMPL 3
.Os
.Sh NAME
.Nm http_servlet_tmpl
.Nd HTTP servlet for template files
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In stdio.h
.In netinet/in.h
.In openssl/ssl.h
.In pdel/tmpl/tmpl.h
.In pdel/http/http_defs.h
.In pdel/http/http_server.h
.In pdel/http/servlet/tmpl.h
.Ft "struct http_servlet *"
.Fn http_servlet_tmpl_create "struct http_servlet_tmpl_info *info"
.Vt extern tmpl_handler_t http_servlet_tmpl_func_query ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_query_exists ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_query_string ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_get_header ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_set_header ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_remove_header ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_redirect ;
.Vt extern tmpl_handler_t http_servlet_tmpl_func_unbuffer ;
.Sh DESCRIPTION
.Fn http_servlet_tmpl_create
creates a new servlet that serves HTTP requests by executing a
.Xr tmpl 3
template file.
The output of the template is sent as the body of the HTTP response.
By default, the template output is buffered in memory and after
template execution has finished the HTTP response is sent.
.Pp
.Fa info
is a pointer to a
.Li "struct http_servlet_tmpl_info" ,
which contains a
.Li "struct http_servlet_tmpl_tinfo" :
.Pp
.Bd -literal -compact -offset 3n
/* Function to free user 'arg' */
typedef void	http_servlet_tmpl_free_t(void *arg);

/* Information required by tmpl(3) library */
struct http_servlet_tmpl_tinfo {
    int                      flags;          /* tmpl_execute() flags */
    const char               *mtype;         /* tmpl string mem type */
    tmpl_handler_t           *handler;       /* tmpl function handler */
    tmpl_errfmtr_t           *errfmtr;       /* tmpl error formatter */
    void                     *arg;           /* opaque argument */
    http_servlet_tmpl_free_t *freer;         /* destructor for 'arg' */
};

/* Information required for the tmpl servlet */
struct http_servlet_tmpl_info {
    const char               *path;          /* template file */
    const char               *mime_type;     /* default mime type */
    const char               *mime_encoding; /* default mime encoding */
    http_logger_t            *logger;        /* http error logger */
    struct http_servlet_tmpl_tinfo   tinfo;  /* info for tmpl(3) */
};
.Ed
.Pp
.Fa path
is the pathname of the template file.
The file is parsed when the servlet is first executed, and the parsed
template is cached for use by subsequent servlet invocations.
If the file modification timestamp changes, the file is parsed again.
.Pp
.Fa mime_type
and
.Fa mime_encoding
specify the default MIME type and encoding for the template output.
The template itself may change the MIME type however.
If
.Fa mime_type
is
.Dv NULL ,
"text/html; charset=iso-8859-1" is used.
.Pp
The
.Fa logger
is a logging function whose type is defined in
.Xr http_server 3 .
.Pp
The
.Fa tinfo
structure provides information required by the
.Xr tmpl 3
library in order to create and execute the template.
The
.Fa mtype ,
.Fa handler ,
and
.Fa errfmtr
fields are passed to
.Fn tmpl_ctx_create ;
The
.Fa flags
field is passed to
.Fn tmpl_execute .
.Pp
The
.Fa arg
field is copied into a
.Li "struct http_servlet_tmpl_arg" ,
and a pointer to this structure is used as the user cookie field
.Fa arg
passed to
.Fn tmpl_ctx_create :
.Pp
.Bd -literal -compact -offset 3n
struct http_servlet_tmpl_arg {
    void                     *arg;   /* arg from 'tinfo' */
    struct http_request      *req;   /* http request */
    struct http_response     *resp;  /* http response */
};
.Ed
.Pp
Therefore, by casting the pointer returned from
.Fn tmpl_ctx_get_arg
to a
.Li "struct http_servlet_tmpl_arg *" ,
.Fn handler
can access both the original
.Fa arg
as well as the HTTP request and response objects.
.Pp
For an HTTP POST with MIME type
.Dq application/x-www-form-urlencoded
the servlet will automatically read in the URL-encoded name, value pairs,
making them accessible via
.Xr http_request_get_value 3
(see also
.Fn http_servlet_tmpl_func_query
below).
.Pp
When the servlet is destroyed, if the
.Fa freer
field is not
.Dv NULL ,
it is invoked to release resources associated with
.Fa arg .
.\"
.Ss Built-in template functions
.\"
The
.Nm http_servlet_tmpl
servlet includes the following built-in
.Xr tmpl 3
user functions.
These functions all assume the
.Xr tmpl 3
user cookie is a
.Li "struct http_servlet_tmpl_arg *".
.Pp
.Bl -hang -offset 3n -width 3n
.It Fn http_servlet_tmpl_func_query
This function takes exactly one argument, which is the name of a field
in an HTML form, and returns the value of that field as submitted by
the remote client.
This function works for both GET and POST form submissions.
If no such field was submitted, the empty string is returned.
.It Fn http_servlet_tmpl_func_query_exists
This function takes exactly one argument, which is the name of a field
in an HTML form, and returns "1" or "0" depending on whether a field
having that name was submitted by the client.
.It Fn http_servlet_tmpl_func_query_string
Takes zero arguments. Returns the HTTP query string.
.It Fn http_servlet_tmpl_func_get_header
Takes one argument, and returns the contents of the HTTP header having
that name.
.It Fn http_servlet_tmpl_func_set_header
Takes two arguments, the name of an HTTP header and its value, and
sets the corresponding HTTP response header.
This function has no effect if the headers have already been sent.
.It Fn http_servlet_tmpl_func_remove_header
Takes one argument, the name of an HTTP header, and removes the
corresponding HTTP response header.
This function has no effect if the headers have already been sent.
.It Fn http_servlet_tmpl_func_redirect
Takes one argument, which must be a valid URL.
Forces an HTTP redirect response to the URL.
This function only works if the servlet has not created any output yet.
.It Fn http_servlet_tmpl_func_unbuffer
Takes zero arguments.
Unbuffers the servlet output, so that the output is written
directly to the network instead of first into a memory buffer. 
This is done by calling
.Xr http_response_send_headers 3
with a non-zero
.Fa unbuffer
argument.
This function should be called in servlets that produce a high
volume of output.
You can't modify the HTTP response headers once this has been called. 
.El
.Pp
The
.Nm http_servlet_tmpl
servlet by default buffers the entire template output 
to facilitate HTTP keep-alive.
For templates that generate voluminous output, this could consume
excessive memory.
The solution is to implement a template function using
.Fn http_servlet_tmpl_func_unbuffer
and invoke this function within those templates.
.Sh IMPLEMENTATION NOTES
Multiple instances of the same
.Nm http_servlet_tmpl
servlet may be executing at the same time.
Therefore, any user-supplied template functions called must
be thread-safe.
.Pp
Since it's running as a servlet, the thread executing
.Fn handler
and
.Fn errfmtr
may be canceled at any cancellation point.
Therefore, these functions should be written so as to not leak
resources if this happens.
.Sh RETURN VALUES
On failure,
.Fn http_servlet_tmpl_create
returns
.Dv NULL
and sets
.Va errno
to an appropriate value.
.Pp
The built-in
.Xr tmpl 3
user functions return
.Dv NULL
with
.Va errno
set to
.Er EINVAL
if the wrong number of arguments is passed.
They may also return
.Dv NULL
with
.Va errno
set as a result of other system errors.
.Sh SEE ALSO
.Xr http_request 3 ,
.Xr http_response 3 ,
.Xr http_server 3 ,
.Xr http_servlet 3 ,
.Xr libpdel 3 ,
.Xr tmpl 3
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/http/servlet/http_servlet_tmpl.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <netinet/in_systm.h>
#include <netinet/in.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <syslog.h>
#include <errno.h>
#include <assert.h>
#include <pthread.h>

#include <openssl/ssl.h>

#include "structs/structs.h"
#include "structs/type/array.h"

#include "tmpl/tmpl.h"
#include "http/http_defs.h"
#include "http/http_server.h"
#include "http/http_servlet.h"
#include "http/servlet/tmpl.h"
#include "util/typed_mem.h"

#define MEM_TYPE	"http_servlet_tmpl"

/* Template servlet private context */
struct tmpl_private {
	struct http_servlet_tmpl_info	info;	/* static template info */
	struct tmpl			*tmpl;	/* parsed template */
	struct timespec			mtime;	/* last mod time of file */
	pthread_rwlock_t		lock;	/* servlet r/w lock */
};

/* Private info per servlet instantiation */
struct tmpl_instance {
	struct tmpl_private		*priv;	/* back pointer to servlet */
	struct tmpl_ctx			*ctx;	/* tmpl exectution context */
	struct http_servlet_tmpl_arg	targ;	/* arg for handler funcs */
};

/* Internal functions */
static http_servlet_run_t	http_servlet_tmpl_run;
static http_servlet_destroy_t	http_servlet_tmpl_destroy;

static void	http_servlet_tmpl_run_cleanup(void *arg);

/*
 * Create a new template servlet.
 */
struct http_servlet *
http_servlet_tmpl_create(struct http_servlet_tmpl_info *info)
{
	struct http_servlet *servlet = NULL;
	struct tmpl_private *priv = NULL;

	/* Create servlet */
	if ((servlet = MALLOC(MEM_TYPE, sizeof(*servlet))) == NULL)
		goto fail;
	memset(servlet, 0, sizeof(*servlet));
	servlet->run = http_servlet_tmpl_run;
	servlet->destroy = http_servlet_tmpl_destroy;

	/* Initialize private info */
	if ((priv = MALLOC(MEM_TYPE, sizeof(*priv))) == NULL)
		goto fail;
	memset(priv, 0, sizeof(*priv));
	if (info->path != NULL
	    && (priv->info.path = STRDUP(MEM_TYPE, info->path)) == NULL)
		goto fail;
	if (info->mime_type != NULL
	    && (priv->info.mime_type = STRDUP(MEM_TYPE,
	      info->mime_type)) == NULL)
		goto fail;
	if (info->mime_encoding != NULL
	    && (priv->info.mime_encoding = STRDUP(MEM_TYPE,
	      info->mime_encoding)) == NULL)
		goto fail;
	priv->info.logger = info->logger;
	if (_http_servlet_tmpl_copy_tinfo(&priv->info.tinfo,
	    &info->tinfo) == -1)
		goto fail;
	pthread_rwlock_init(&priv->lock, NULL);
	servlet->arg = priv;

	/* OK */
	return (servlet);

fail:
	if (priv != NULL) {
		FREE(MEM_TYPE, (char *)priv->info.path);
		FREE(MEM_TYPE, (char *)priv->info.mime_type);
		FREE(MEM_TYPE, (char *)priv->info.mime_encoding);
		_http_servlet_tmpl_free_tinfo(&priv->info.tinfo);
		FREE(MEM_TYPE, priv);
	}
	if (servlet != NULL)
		FREE(MEM_TYPE, servlet);
	return (NULL);
}

/*
 * Destroy template servlet.
 */
static void
http_servlet_tmpl_destroy(struct http_servlet *servlet)
{
	struct tmpl_private *const priv = servlet->arg;
	struct http_servlet_tmpl_info *const info = &priv->info;
	int r;

	/* Acquire the lock to avoid race condition */
	r = pthread_rwlock_wrlock(&priv->lock);
	assert(r == 0);

	/* Destroy user argument */
	if (info->tinfo.freer != NULL)
		(*info->tinfo.freer)(info->tinfo.arg);

	/* Destroy template context */
	tmpl_destroy(&priv->tmpl);

	/* Release the lock */
	r = pthread_rwlock_unlock(&priv->lock);
	assert(r == 0);

	/* Free structure */
	FREE(MEM_TYPE, (char *)info->path);
	FREE(MEM_TYPE, (char *)info->mime_type);
	FREE(MEM_TYPE, (char *)info->mime_encoding);
	_http_servlet_tmpl_free_tinfo(&info->tinfo);
	FREE(MEM_TYPE, priv);
	FREE(MEM_TYPE, servlet);
}

/*
 * Run template servlet
 */
static int
http_servlet_tmpl_run(struct http_servlet *servlet,
	struct http_request *req, struct http_response *resp)
{
	struct tmpl_private *const priv = servlet->arg;
	struct http_servlet_tmpl_info *const info = &priv->info;
	struct http_servlet_tmpl_tinfo *const tinfo = &priv->info.tinfo;
	struct tmpl_instance *this = NULL;
	FILE *output = NULL;
	const char *hval;
	struct stat sb;
	int num_errors;
	int r;

	/* Construct per-instance state */
	if ((this = MALLOC(MEM_TYPE, sizeof(*this))) == NULL) {
		(*info->logger)(LOG_ERR, "%s: %s: %s",
		    __FUNCTION__, "malloc", strerror(errno));
		return (-1);
	}
	memset(this, 0, sizeof(*this));
	this->priv = priv;

	/* Grab lock to avoid race with http_servlet_tmpl_destroy() */
	r = pthread_rwlock_rdlock(&priv->lock);
	assert(r == 0);

	/* Push cleanup hook in case thread gets canceled */
	pthread_cleanup_push(http_servlet_tmpl_run_cleanup, this);

	/* Get servlet output stream (buffered) */
	if ((output = http_response_get_output(resp, 1)) == NULL) {
		(*info->logger)(LOG_ERR, "can't get template output: %s",
		    strerror(errno));
		goto fail_errno;
	}

	/* Set MIME type */
	if (info->mime_type == NULL) {
		http_response_set_header(resp, 0,
		    HTTP_HEADER_CONTENT_TYPE, "text/html; charset=iso-8859-1");
	} else {
		http_response_set_header(resp, 0,
		    HTTP_HEADER_CONTENT_TYPE, "%s", info->mime_type);
		if (info->mime_encoding != NULL) {
			http_response_set_header(resp,
			0, HTTP_HEADER_CONTENT_ENCODING,
			    "%s", info->mime_encoding);
		}
	}

	/* Assume servlet output is not cachable */
	http_response_set_header(resp, 1, HTTP_HEADER_PRAGMA, "no-cache");
	http_response_set_header(resp, 0,
	    HTTP_HEADER_CACHE_CONTROL, "no-cache");

	/* Get modification timestamp of the template file */
	if (stat(info->path, &sb) == -1) {
		(*info->logger)(LOG_ERR, "%s: %s: %s",
		    __FUNCTION__, info->path, strerror(errno));
		memset(&sb.st_mtime, 0, sizeof(sb.st_mtime));
	}

	/* Invalidate cached template if template file has changed */
	if (priv->tmpl != NULL
	    && memcmp(&sb.st_mtime, &priv->mtime, sizeof(priv->mtime)) != 0) {
		(*info->logger)(LOG_INFO,
		    "template \"%s\" was updated", info->path);
		tmpl_destroy(&priv->tmpl);
	}

	/* Do we need to (re)parse the template? */
	if (priv->tmpl == NULL) {

		/* Parse template file */
		if ((priv->tmpl = tmpl_create_mmap(info->path,
		    &num_errors, tinfo->mtype)) == NULL) {
			(*info->logger)(LOG_ERR,
			    "can't create template from \"%s\": %s",
			    info->path, strerror(errno));
			goto fail_errno;
		}

		/* Check for an error from tmpl_create() */
		if (priv->tmpl == NULL) {
			(*info->logger)(LOG_ERR,
			    "can't create \"%s\" template: %s", info->path,
			    strerror(errno));
			goto fail_errno;
		}

		/* Warn if there were any parse errors */
		if (num_errors != 0) {
			(*info->logger)(LOG_WARNING,
			    "%d parse error%s in template \"%s\"",
			    num_errors, num_errors == 1 ? "" : "s",
			    info->path);
		}

		/* Update last modified time */
		memcpy(&priv->mtime, &sb.st_mtime, sizeof(priv->mtime));
	}

	/* Read URL-encoded form data if this is a normal POST */
	if (strcmp(http_request_get_method(req), HTTP_METHOD_POST) == 0
	    && (hval = http_request_get_header(req,
	      HTTP_HEADER_CONTENT_TYPE)) != NULL
	    && strcasecmp(hval, HTTP_CTYPE_FORM_URLENCODED) == 0) {
		if (http_request_read_url_encoded_values(req) == -1) {
			(*info->logger)(LOG_ERR,
			    "error reading %s data for \"%s\" template: %s",
			    HTTP_METHOD_POST, info->path, strerror(errno));
			goto fail_errno;
		}
	}

	/* Fill in handler function cookie */
	this->targ.arg = info->tinfo.arg;
	this->targ.req = req;
	this->targ.resp = resp;

	/* Create tmpl execution context */
	if ((this->ctx = tmpl_ctx_create(&this->targ,
	    tinfo->mtype, tinfo->handler, tinfo->errfmtr)) == NULL) {
		(*info->logger)(LOG_ERR, "%s: %s: %s",
		    __FUNCTION__, "tmpl_ctx_create", strerror(errno));
		goto fail_errno;
	}

	/* Execute template */
	if (tmpl_execute(priv->tmpl, this->ctx, output, tinfo->flags) == -1) {
		(*info->logger)(LOG_ERR, "can't execute \"%s\" template: %s",
		    info->path, strerror(errno));
		goto fail_errno;
	}

	/* OK */
	goto done;

fail_errno:
	/* Fail with appropriate error response */
	http_response_send_errno_error(resp);

done:
	/* Done */
	if (output != NULL)
		fclose(output);
	pthread_cleanup_pop(1);
	return (1);
}

/*
 * Cleanup after http_servlet_tmpl_run().
 */
static void
http_servlet_tmpl_run_cleanup(void *arg)
{
	struct tmpl_instance *const this = arg;
	int r;

	if (this->ctx != NULL)
		tmpl_ctx_destroy(&this->ctx);
	r = pthread_rwlock_unlock(&this->priv->lock);
	assert(r == 0);
	FREE(MEM_TYPE, this);
}

/*
 * Copy a 'tinfo' structure.
 */
int
_http_servlet_tmpl_copy_tinfo(struct http_servlet_tmpl_tinfo *dst,
	const struct http_servlet_tmpl_tinfo *src)
{

	/* Copy stuff */
	memset(dst, 0, sizeof(*dst));
	dst->flags = src->flags;
	dst->handler = src->handler;
	dst->errfmtr = src->errfmtr;
	if (src->mtype != NULL
	    && (dst->mtype = STRDUP(MEM_TYPE, src->mtype)) == NULL)
		goto fail;
	dst->arg = src->arg;
	dst->freer = src->freer;

	/* Done */
	return (0);

fail:
	/* Clean up after failure */
	FREE(MEM_TYPE, (char *)dst->mtype);
	memset(dst, 0, sizeof(*dst));
	return (-1);
}

/*
 * Free a copied 'tinfo' structure.
 */
void
_http_servlet_tmpl_free_tinfo(struct http_servlet_tmpl_tinfo *tinfo)
{
	FREE(MEM_TYPE, (char *)tinfo->mtype);
	memset(tinfo, 0, sizeof(*tinfo));
}

/************************************************************************
			    TMPL USER FUNCTIONS
************************************************************************/

char *
http_servlet_tmpl_func_query(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);
	const char *value;

	if (ac != 2) {
		errno = EINVAL;
		return (NULL);
	}
	if ((value = http_request_get_value(arg->req, av[1], 0)) == NULL)
		value = "";
	return (STRDUP(mtype, value));
}

char *
http_servlet_tmpl_func_query_exists(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);
	char buf[2];

	if (ac != 2) {
		errno = EINVAL;
		return (NULL);
	}
	snprintf(buf, sizeof(buf), "%d",
	    http_request_get_value(arg->req, av[1], 0) != NULL);
	return (STRDUP(mtype, buf));
}

char *
http_servlet_tmpl_func_query_string(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);
	const char *eqs = http_request_get_query_string(arg->req);
	char *dqs;

	if (ac != 1) {
		errno = EINVAL;
		return (NULL);
	}
	/* URL-decode query string */
	if ((dqs = MALLOC(mtype, strlen(eqs) + 1)) == NULL)
		return (NULL);
	http_request_url_decode(eqs, dqs);

	/* Return it */
	return (dqs);
}

char *
http_servlet_tmpl_func_get_header(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);
	const char *hval;

	if (ac != 2) {
		errno = EINVAL;
		return (NULL);
	}
	if ((hval = http_request_get_header(arg->req, av[1])) == NULL)
		hval = "";
	return (STRDUP(mtype, hval));
}

char *
http_servlet_tmpl_func_set_header(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);

	if (ac != 3) {
		errno = EINVAL;
		return (NULL);
	}
	if (http_response_set_header(arg->resp, 0, av[1], "%s", av[2]) == -1)
		return (NULL);
	return (STRDUP(mtype, ""));
}

char *
http_servlet_tmpl_func_remove_header(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);

	if (ac != 2) {
		errno = EINVAL;
		return (NULL);
	}
	(void)http_response_remove_header(arg->resp, av[1]);
	return (STRDUP(mtype, ""));
}

char *
http_servlet_tmpl_func_unbuffer(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);

	if (ac != 1) {
		errno = EINVAL;
		return (NULL);
	}
	http_response_send_headers(arg->resp, 1);
	return (STRDUP(mtype, ""));
}

char *
http_servlet_tmpl_func_redirect(struct tmpl_ctx *ctx,
	char **errmsgp, int ac, char **av)
{
	struct http_servlet_tmpl_arg *const arg = tmpl_ctx_get_arg(ctx);
	const char *const mtype = tmpl_ctx_get_mtype(ctx);

	if (ac != 2) {
		errno = EINVAL;
		return (NULL);
	}
	if (http_response_set_header(arg->resp, 0,
	    HTTP_HEADER_LOCATION, "%s", av[1]) == -1)
		return (NULL);
	http_response_send_error(arg->resp, HTTP_STATUS_FOUND, NULL);
	return (STRDUP(mtype, ""));
}

